From 5d620970d71250284f4cb40e121423c320915f17 Mon Sep 17 00:00:00 2001
From: tm <tm@x>
Date: Sun, 18 Oct 2026 18:55:36 +0100
Subject: [PATCH] > <= >= of two variables or constants compiled as
 instructions

They were calls of functions (system.lsp; in TeXmacs functions with any
number of arguments), at each step of the loops which compare characters,
3 to 4 times slower than < . compile-app (simple-compare) compiles them
with < = number? when their two arguments are variables or constants (no
order of evaluation to keep) and none of these names is bound locally or
by resolve-global; the functions stay for the other calls and as values.

<= and >= follow the order of < on any values, but are false when a
number is a NaN: (<= 1 +nan.0) was true. system.lsp defines them so too.
---
 compiler.lsp       |  33 ++++
 flisp.boot         | 444 +++++++++++++++++++++++----------------------
 system.lsp         |   8 +-
 tests/unittest.lsp |   7 +
 4 files changed, 272 insertions(+), 220 deletions(-)

diff --git a/compiler.lsp b/compiler.lsp
index 9244206..8b50146 100644
--- a/compiler.lsp
+++ b/compiler.lsp
@@ -401,7 +401,40 @@
 		    (emit g (if tail? 'tapply 'apply) nargs)))
       (else      (emit g b)))))
 
+; TeXmacs: > <= >= of two arguments which are variables or constants, as
+; the instructions < = number? (see also compile-app), as system.lsp
+; defines them; they were calls of functions, at each step of the loops
+; which compare characters. The arguments are evaluated in any order then,
+; and the function stays for the other calls and as a value.
+(define (simple-compare x env)
+  (let ((head (car x)))
+    (and (memq head '(> <= >=))
+	 (length= x 3)
+	 (not (in-env? head env))
+	 (eq? (resolve-global head) head)
+	 (let loop ((l '(< = number?)))
+	   (or (null? l)
+	       (and (not (in-env? (car l) env))
+		    (eq? (resolve-global (car l)) (car l))
+		    (loop (cdr l)))))
+	 (let ((simple? (lambda (a) (or (atom? a) (eq? (car a) 'quote))))
+	       (a (cadr x)) (b (caddr x)))
+	   (and (simple? a) (simple? b)
+		(case head
+		  ((>)  (list '< b a))
+		  ((<=) (compare<= a b))
+		  (else (compare<= b a))))))))
+
+; (<= a b): the order of < (any values), but false with a NaN
+(define (compare<= a b)
+  `(if (< ,b ,a) #f (if (number? ,a) (if (< ,a ,b) #t (= ,a ,b)) #t)))
+
 (define (compile-app g env tail? x)
+  (let ((c (simple-compare x env)))
+    (if c (compile-in g env tail? c)
+	(compile-app* g env tail? x))))
+
+(define (compile-app* g env tail? x)
   (let* ((head  (car x))
 	 (ghead (if (and (symbol? head) (not (in-env? head env)))
 		    (resolve-global head)
diff --git a/flisp.boot b/flisp.boot
index 4b3ddf6..459e260 100644
--- a/flisp.boot
+++ b/flisp.boot
@@ -1,8 +1,8 @@
 (%fl:1+ #3=#fn("7000r1|aw;" #() 1+) %fl:1-
 	#4=#fn("7000r1|ax;" #() 1-) %fl:1arg-lambda? #5=#fn("8000r1|F16Z02|Mc0<16P02|NF16H02e1|31F16=02e2e1|31a42;" #(lambda
   %fl:cadr %fl:length=) 1arg-lambda?)
-	%fl:<= #6=#fn("7000r2}|X17B02e0|3116802e0}31@;" #(%fl:nan?) <=) %fl:>
-	#7=#fn("7000r2}|X;" #() >) %fl:>= #8=#fn("7000r2|}X17B02e0|3116802e0}31@;" #(%fl:nan?) >=)
+	%fl:<= #6=#fn("7000r2}|X640^;|D6>0|}X640];|}W;];" #() <=) %fl:>
+	#7=#fn("7000r2}|X;" #() >) %fl:>= #8=#fn("8000r2e0}|42;" #(%fl:<=) >=)
 	%fl:__init_globals #9=#fn("7000r0e0c1<17B02e0c2<17802e0c3<6>0c4k52c6k75;0c8k52c9k72e:k;2e<k=2e>k?;" #(*os-name*
   win32 win64 windows "\\" *directory-separator* "\x0d\n" *linefeed* "/" "\n"
   *stdout* *output-stream* *stdin* *input-stream* *stderr* *error-stream*) __init_globals)
@@ -70,29 +70,34 @@
 	%fl:cdddr #55=#fn("6000r1|NNN;" #() cdddr) %fl:cddr
 	#56=#fn("6000r1|NN;" #() cddr) %fl:char? #57=#fn("7000r1c0|31c1<;" #(#fn(typeof)
   wchar) char?)
-	%fl:closure? #58=#fn("7000r1|J16602|G@;" #() closure?) %fl:compile
-	#59=#fn("8000r1e0_|42;" #(%fl:compile-f) compile) %fl:compile-and #60=#fn("<000r4e0|}g2g3]c146;" #(%fl:compile-short-circuit
-  brf) compile-and)
-	%fl:compile-app #61=#fn("7000r4c0qg3M41;" #(#fn("9000r1c0q|C16:02e1|\x7f32@6:0e2|31530|41;" #(#fn("9000r1c0q~C16X02e1~i1132@16J02|E16C02c2|3116902c3|31G6:0c3|31530~41;" #(#fn("<000r1|c0<17702|c1<16=02e2i23b332@6R0e3i20i21i22|c0‚70c4540c5i23NK44;e6i23Nc7326S0e3i20i21^|342c8qe9i20i21i23N3341;c:q|G16J02e;|c<i23N313216802e=|3141;" #(#.<
+	%fl:closure? #58=#fn("7000r1|J16602|G@;" #() closure?) %fl:compare<=
+	#59=#fn("@000r2c0c1}|L3^c0c2|L2c0c1|}L3]c3|}L3L4]L4L4;" #(if < number?
+								  =) compare<=)
+	%fl:compile #60=#fn("8000r1e0_|42;" #(%fl:compile-f) compile)
+	%fl:compile-and #61=#fn("<000r4e0|}g2g3]c146;" #(%fl:compile-short-circuit
+							 brf) compile-and)
+	%fl:compile-app #62=#fn("9000r4c0qe1g3}3241;" #(#fn(":000r1|6=0e0~\x7fi02|44;e1~\x7fi02i0344;" #(%fl:compile-in
+  %fl:compile-app*)) %fl:simple-compare) compile-app)
+	%fl:compile-app* #63=#fn("7000r4c0qg3M41;" #(#fn("9000r1c0q|C16:02e1|\x7f32@6:0e2|31530|41;" #(#fn("9000r1c0q~C16X02e1~i1132@16J02|E16C02c2|3116902c3|31G6:0c3|31530~41;" #(#fn("<000r1|c0<17702|c1<16=02e2i23b332@6R0e3i20i21i22|c0‚70c4540c5i23NK44;e6i23Nc7326S0e3i20i21^|342c8qe9i20i21i23N3341;c:q|G16J02e;|c<i23N313216802e=|3141;" #(#.<
   #.= %fl:length= %fl:compile-in nary< nary= %fl:length> 255 #fn(":000r1e0i30i32670c1540c2|43;" #(%fl:emit
   tcall.l call.l)) %fl:compile-arglist #fn(";000r1~c0<16c02i10c0<16X02e1~i3132@16J02c2c031e3>16<02e4i33b2326O0e5i30i31^e3i3331342e6i30c042;|7A0e5i30i31^~34540c72c8qe9i30i31i33N3341;" #(cadr
   %fl:in-env? #fn(top-level-value) %fl:cadr %fl:length= %fl:compile-in %fl:emit
   #{#<unspecified>}# #fn("=000r1~6H0e0i40i41i42i43i10~|47;e1i40i42670c2540c3|43;" #(%fl:compile-builtin-call
   %fl:emit tcall call)) %fl:compile-arglist))
   %fl:builtin-argc-ok? #fn(length) %fl:builtin->instruction)) %fl:in-env? #fn(constant?)
-  #fn(top-level-value))) %fl:in-env? resolve-global))) compile-app)
-	%fl:compile-arglist #62=#fn("8000r3e0c1qg2322c2g241;" #(%fl:for-each #fn(":000r1e0~\x7f^|44;" #(%fl:compile-in))
+  #fn(top-level-value))) %fl:in-env? resolve-global))) compile-app*)
+	%fl:compile-arglist #64=#fn("8000r3e0c1qg2322c2g241;" #(%fl:for-each #fn(":000r1e0~\x7f^|44;" #(%fl:compile-in))
 								#fn(length)) compile-arglist)
-	%fl:compile-begin #63=#fn(":000r4g3?6?0e0|}g2e13044;g3N?6>0e0|}g2g3M44;e0|}^g3M342e2|c3322e4|}g2g3N44;" #(%fl:compile-in
+	%fl:compile-begin #65=#fn(":000r4g3?6?0e0|}g2e13044;g3N?6>0e0|}g2g3M44;e0|}^g3M342e2|c3322e4|}g2g3N44;" #(%fl:compile-in
   %fl:void %fl:emit pop %fl:compile-begin) compile-begin)
-	%fl:compile-builtin-call #64=#fn(":000r7c0qc1e2g4^3341;" #(#fn("8000r1|16=02e0i03N|32@6=0e1i05|32540c22c3qi0541;" #(%fl:length=
+	%fl:compile-builtin-call #66=#fn(":000r7c0qc1e2g4^3341;" #(#fn("8000r1|16=02e0i03N|32@6=0e1i05|32540c22c3qi0541;" #(%fl:length=
   %fl:argc-error #{#<unspecified>}# #fn(":000r1|c0‚R0i16`W6<0e1i10c242;e1i10i15i1643;|c3‚e0i16`W6<0e1i10c442;i16b2W6<0e1i10c542;e1i10i15i1643;|c6‚v0i16`W6;0e7i15a42;i16aW6<0e1i10c842;i16b2W6<0e1i10c942;e1i10i15i1643;|c:‚R0i16`W6<0e1i10c;42;e1i10i15i1643;|c<‚Q0i16`W6;0e7i15a42;e1i10i15i1643;|c=‚T0i16`W6>0e1i10c>c?43;e1i10i15i1643;|c@‚]0i16b2X6<0e7i15b242;e1i10i12670cA540c@i1643;e1i10i1542;" #(list
   %fl:emit loadnil + load0 add2 - %fl:argc-error neg sub2 * load1 / vector
   loadv #() apply tapply)))) #fn(get) arg-counts) compile-builtin-call)
-	%fl:compile-f #65=#fn("8000r2e0c1qc242;" #(%fl:call-with-values
+	%fl:compile-f #67=#fn("8000r2e0c1qc242;" #(%fl:call-with-values
 						   #fn("8000r0e0~\x7f42;" #(%fl:compile-f-))
 						   #fn("6000r2|;" #())) compile-f)
-	%fl:compile-f- #66=#fn("8000r2c0qc1c142;" #(#fn(">000r2c0qm02c1qm12c2qe330e4\x7f31e5e4\x7f3131e6e4\x7f3131e7c8e4\x7f3132e5\x7f31i10‚70c9570e5\x7f3146;" #(#fn("9000r1c0qe1|31F6N0e2|31F6=0c3e1|31K570e4|31560e53041;" #(#fn("8000r1c0qe1|3141;" #(#fn(":000r1|…40~;c0c1|~i4034c2c3|32K;" #(#fn(list*)
+	%fl:compile-f- #68=#fn("8000r2c0qc1c142;" #(#fn(">000r2c0qm02c1qm12c2qe330e4\x7f31e5e4\x7f3131e6e4\x7f3131e7c8e4\x7f3132e5\x7f31i10‚70c9570e5\x7f3146;" #(#fn("9000r1c0qe1|31F6N0e2|31F6=0c3e1|31K570e4|31560e53041;" #(#fn("8000r1c0qe1|3141;" #(#fn(":000r1|…40~;c0c1|~i4034c2c3|32K;" #(#fn(list*)
   lambda #fn(map) #fn("6000r1e040;" #(%fl:void))))
   %fl:get-defined-vars)) %fl:cddr %fl:cdddr begin %fl:caddr %fl:void) lambda-body)
   #fn("7000r1e0|31i20‚80e1|41;~|41;" #(%fl:lastcdr %fl:caddr) lam:body)
@@ -106,12 +111,12 @@
   %fl:keyword-arg?)) #fn(length))) #fn(length)))
   %fl:make-code-emitter %fl:cadr %fl:lastcdr %fl:lambda-vars %fl:filter #.pair?
   lambda)) #{#<unspecified>}#) #0=#(%defines-processed% ()))
-	%fl:compile-for #67=#fn(":000r5e0g4316X0e1|}^g2342e1|}^g3342e1|}^g4342e2|c342;e4c541;" #(%fl:1arg-lambda?
+	%fl:compile-for #69=#fn(":000r5e0g4316X0e1|}^g2342e1|}^g3342e1|}^g4342e2|c342;e4c541;" #(%fl:1arg-lambda?
   %fl:compile-in %fl:emit for %fl:error "for: third form must be a 1-argument lambda") compile-for)
-	%fl:compile-if #68=#fn("<000r4c0qe1|31e1|31e2g331e3g331e4g331F6;0e5g331560e63045;" #(#fn(";000r5g2]‚>0e0~\x7fi02g344;g2^‚>0e0~\x7fi02g444;e0~\x7f^g2342e1~c2|332e0~\x7fi02g3342i026<0e1~c3325:0e1~c4}332e5~|322e0~\x7fi02g4342e5~}42;" #(%fl:compile-in
+	%fl:compile-if #70=#fn("<000r4c0qe1|31e1|31e2g331e3g331e4g331F6;0e5g331560e63045;" #(#fn(";000r5g2]‚>0e0~\x7fi02g344;g2^‚>0e0~\x7fi02g444;e0~\x7f^g2342e1~c2|332e0~\x7fi02g3342i026<0e1~c3325:0e1~c4}332e5~|322e0~\x7fi02g4342e5~}42;" #(%fl:compile-in
   %fl:emit brf ret jmp %fl:mark-label)) %fl:make-label %fl:cadr %fl:caddr
   %fl:cdddr %fl:cadddr %fl:void) compile-if)
-	%fl:compile-in #69=#fn(";000r4g3C6=0e0|}g3c144;g3?6¯0g3`‚:0e2|c342;g3a‚:0e2|c442;g3]‚:0e2|c542;g3^‚:0e2|c642;g3_‚:0e2|c742;e8g3316<0e2|c9g343;c:g3316C0e;|}g2c<c=31L144;e2|c>g343;g3MC@17W02c?g3Me@32@16H02eAg3M31E17;02eBg3M}326=0eC|}g2g344;cDqc?g3Me@32@16:02eEg3}3241;" #(%fl:compile-sym
+	%fl:compile-in #71=#fn(";000r4g3C6=0e0|}g3c144;g3?6¯0g3`‚:0e2|c342;g3a‚:0e2|c442;g3]‚:0e2|c542;g3^‚:0e2|c642;g3_‚:0e2|c742;e8g3316<0e2|c9g343;c:g3316C0e;|}g2c<c=31L144;e2|c>g343;g3MC@17W02c?g3Me@32@16H02eAg3M31E17;02eBg3M}326=0eC|}g2g344;cDqc?g3Me@32@16:02eEg3}3241;" #(%fl:compile-sym
   #(loada loadc loadg) %fl:emit load0 load1 loadt loadf loadnil %fl:fits-i8
   loadi8 #fn(eof-object?) %fl:compile-in #fn(top-level-value) eof-object loadv
   #fn(memq) special-forms resolve-global %fl:in-env? %fl:compile-app #fn(":000r1|6=0e0~\x7fi02|44;c1qi03M41;" #(%fl:compile-in
@@ -128,35 +133,35 @@
   %fl:compile-sym #(seta setc setg) define %fl:expand-define trycatch
   %fl:1arg-lambda? "trycatch: second form must be a 1-argument lambda"
   %fl:compile-app)))) compile-unknown-call) compile-in)
-	%fl:compile-or #70=#fn("<000r4e0|}g2g3^c146;" #(%fl:compile-short-circuit
+	%fl:compile-or #72=#fn("<000r4e0|}g2g3^c146;" #(%fl:compile-short-circuit
 							brt) compile-or)
-	%fl:compile-prog1 #71=#fn(";000r3e0|}^e1g231342e2g231F6H0e3|}^e2g231342e4|c542;c6;" #(%fl:compile-in
+	%fl:compile-prog1 #73=#fn(";000r3e0|}^e1g231342e2g231F6H0e3|}^e2g231342e4|c542;c6;" #(%fl:compile-in
   %fl:cadr %fl:cddr %fl:compile-begin %fl:emit pop #{#<unspecified>}#) compile-prog1)
-	%fl:compile-short-circuit #72=#fn(":000r6g3?6=0e0|}g2g444;g3N?6>0e0|}g2g3M44;c1qe2|3141;" #(%fl:compile-in
+	%fl:compile-short-circuit #74=#fn(":000r6g3?6=0e0|}g2g444;g3N?6>0e0|}g2g3M44;c1qe2|3141;" #(%fl:compile-in
   #fn("<000r1e0~\x7f^i03M342e1~c2322e1~i05|332e1~c3322e4~\x7fi02i03Ni04i05362e5~|42;" #(%fl:compile-in
   %fl:emit dup pop %fl:compile-short-circuit %fl:mark-label)) %fl:make-label) compile-short-circuit)
-	%fl:compile-sym #73=#fn(";000r4c0qe1g2}`]3441;" #(#fn(":000r1|D6>0e0~i03`[|43;|MD6R0e0~i03a[|M|N342e1~e2\x7fN31a|MS342;c3qe4i023141;" #(%fl:emit
+	%fl:compile-sym #75=#fn(";000r4c0qe1g2}`]3441;" #(#fn(":000r1|D6>0e0~i03`[|43;|MD6R0e0~i03a[|M|N342e1~e2\x7fN31a|MS342;c3qe4i023141;" #(%fl:emit
   %fl:bcode:cdepth %fl:nnn #fn(":000r1c0|3116<02e1c2|31316A0e3i10c4c2|3143;e3i10i13b2[|43;" #(#fn(constant?)
   %fl:printable? #fn(top-level-value) %fl:emit loadv)) resolve-global))
 							  %fl:lookup-sym) compile-sym)
-	%fl:compile-thunk #74=#fn(";000r1e0c1c2L1_L1|L1~3441;" #(%fl:compile #fn(nconc)
+	%fl:compile-thunk #76=#fn(";000r1e0c1c2L1_L1|L1~3441;" #(%fl:compile #fn(nconc)
 								 lambda) #0#)
-	%fl:compile-while #75=#fn("9000r4c0qe1|31e1|3142;" #(#fn(":000r2e0~\x7f^e130342e2~|322e0~\x7f^i02342e3~c4}332e3~c5322e0~\x7f^i03342e3~c6|332e2~}42;" #(%fl:compile-in
+	%fl:compile-while #77=#fn("9000r4c0qe1|31e1|3142;" #(#fn(":000r2e0~\x7f^e130342e2~|322e0~\x7f^i02342e3~c4}332e3~c5322e0~\x7f^i03342e3~c6|332e2~}42;" #(%fl:compile-in
   %fl:void %fl:mark-label %fl:emit brf pop jmp)) %fl:make-label) compile-while)
-	%fl:const-to-idx-vec #76=#fn("9000r1c0qc1e2|313141;" #(#fn("9000r1e0c1qe2~31322|;" #(%fl:table.foreach
+	%fl:const-to-idx-vec #78=#fn("9000r1c0qc1e2|313141;" #(#fn("9000r1e0c1qe2~31322|;" #(%fl:table.foreach
   #fn("8000r2~}|\\;" #()) %fl:bcode:ctable))
 							       #fn(vector.alloc)
 							       %fl:bcode:nconst) const-to-idx-vec)
-	%fl:copy-tree #77=#fn("8000r1|?640|;e0|M31e0|N31K;" #(%fl:copy-tree) copy-tree)
-	%fl:count #78=#fn("7000r2c0qc141;" #(#fn("9000r1c0qm02|~\x7f`43;" #(#fn(":000r3}…50g2;~|}N|}M31690g2aw540g243;" #() count-)))
+	%fl:copy-tree #79=#fn("8000r1|?640|;e0|M31e0|N31K;" #(%fl:copy-tree) copy-tree)
+	%fl:count #80=#fn("7000r2c0qc141;" #(#fn("9000r1c0qm02|~\x7f`43;" #(#fn(":000r3}…50g2;~|}N|}M31690g2aw540g243;" #() count-)))
 					     #{#<unspecified>}#) count)
-	%fl:define-head-name #79=#fn("7000r1|F690e0|M41;|;" #(%fl:define-head-name) define-head-name)
-	%fl:delete-duplicates #80=#fn("8000r1e0|bD326<0c1qc23041;|?640|;c3|M|N42;" #(%fl:length>
+	%fl:define-head-name #81=#fn("7000r1|F690e0|M41;|;" #(%fl:define-head-name) define-head-name)
+	%fl:delete-duplicates #82=#fn("8000r1e0|bD326<0c1qc23041;|?640|;c3|M|N42;" #(%fl:length>
   #fn("8000r1c0qc131~_42;" #(#fn("6000r1c0qm02|;" #(#fn("9000r2|?680e0}41;c1i10|M32690~|N}42;c2i10|M]332~|N|M}K42;" #(%fl:reverse!
   #fn(has?) #fn(put!))))) #{#<unspecified>}#))
   #fn(table) #fn("8000r2e0|}32680e1}41;|e1}31K;" #(%fl:member
 						   %fl:delete-duplicates))) delete-duplicates)
-	%fl:disassemble #81=#fn("=000s1}…C0e0|`322e1302];540c22c3}Mc4|31c5|3143;" #(%fl:disassemble
+	%fl:disassemble #83=#fn("=000s1}…C0e0|`322e1302];540c22c3}Mc4|31c5|3143;" #(%fl:disassemble
   %fl:newline #{#<unspecified>}# #fn("7000r3c0qc141;" #(#fn(":000r1c0qm02`~axc1u2e2c3e4\x7f`32c5332c6qb4c7\x7f3142;" #(#fn("9000r1|J16602|G@6D0e0c1312e2|i10aw42;e3|41;" #(%fl:princ
   "\n" %fl:disassemble %fl:print) print-val)
   #fn("7000r1e0c141;" #(%fl:princ "\t")) %fl:princ "maxstack " %fl:ref-int32-LE
@@ -175,9 +180,9 @@
 						     Instructions))
   #fn(length))) #{#<unspecified>}#)) #fn(function:code)
   #fn(function:vals)) disassemble)
-	%fl:div #82=#fn("8000r2|}V|`X16C02}`X16402a17502b/17402`w;" #() div)
+	%fl:div #84=#fn("8000r2|}V|`X16C02}`X16402a17502b/17402`w;" #() div)
 	%fl:double #fn(double) %fl:emit
-	#83=#fn("G000s2g2…b0}c0<16C02|`[F16:02|`[Mc1<6;0|`[c2O5:0|`}|`[K\\5å0c3}c4326A0e5|g2M32L1m2540c62c7qc8}c932312c:qc8}c;32312}c<‚\\0g2c=>6=0c>m12_m25F0g2c?>6=0c@m12_m2530^540c62}cA‚\\0g2cB>6=0cCm12_m25F0g2cD>6=0cEm12_m2530^540c62cFq|`[F690|`[M530_|`[322|;" #(car
+	#85=#fn("G000s2g2…b0}c0<16C02|`[F16:02|`[Mc1<6;0|`[c2O5:0|`}|`[K\\5å0c3}c4326A0e5|g2M32L1m2540c62c7qc8}c932312c:qc8}c;32312}c<‚\\0g2c=>6=0c>m12_m25F0g2c?>6=0c@m12_m2530^540c62}cA‚\\0g2cB>6=0cCm12_m25F0g2cD>6=0cEm12_m2530^540c62cFq|`[F690|`[M530_|`[322|;" #(car
   cdr cadr #fn(memq) (loadv loadg setg) %fl:bcode:indexfor #{#<unspecified>}#
   #fn("8000r1|16=02e0i02Mc1326;0e2|31o01;c3;" #(%fl:> 255 %fl:cadr
 						#{#<unspecified>}#))
@@ -186,10 +191,10 @@
   255 %fl:cadr #{#<unspecified>}#)) ((loadc loadc.l) (setc setc.l)) loada (0)
   loada0 (1) loada1 loadc (0 0) loadc00 (0 1) loadc01 #fn(">000r2\x7fc0<1602|c1<16;02e2}31c3<6E0~`i02Mc4e5}31KK\\5u0|c1‚B0~`i02Mc6}NKK\\5_0|c7‚B0~`i02Mc8}NKK\\5I0|c3‚B0~`i02Mc9}NKK\\530^17^02\x7fc6<16702|c3<6@0~`i02Mc4}NKK\\;~`e:\x7fi02K}32\\;" #(brf
   not %fl:cadr null? brn %fl:cddr brt eq? brne brnn %fl:nreconc))) emit)
-	%fl:emit-optional-arg-inits #84=#fn("8000r5g2F6=0c0qe1|3141;c2;" #(#fn("<000r1e0~c1i04332e0~c2|332e3~e4i03i0432\x7fK^e5i0231342e0~c6i04332e0~c7322e8~|322e9~\x7fi02Ni03i04aw45;" #(%fl:emit
+	%fl:emit-optional-arg-inits #86=#fn("8000r5g2F6=0c0qe1|3141;c2;" #(#fn("<000r1e0~c1i04332e0~c2|332e3~e4i03i0432\x7fK^e5i0231342e0~c6i04332e0~c7322e8~|322e9~\x7fi02Ni03i04aw45;" #(%fl:emit
   brbound brt %fl:compile-in %fl:list-head %fl:cadar seta pop %fl:mark-label
   %fl:emit-optional-arg-inits)) %fl:make-label #{#<unspecified>}#) emit-optional-arg-inits)
-	%fl:encode-byte-code #85=#fn("8000r1c0e1|3141;" #(#fn("8000r1c0e1|3141;" #(#fn(";000r1c0qe1c2|31b3c2|31b2VT2wc33241;" #(#fn("=000r1c0qc1~31`c230c230c330^^47;" #(#fn("?000r7c0g4c1322c2}|X6ÿ02i10}[m52g5c3‚O0c4g2i10}aw[c5g431332}b2wm15Ï0c0g4e6c7e8~6<0c9qg531540g53231322}awm12}|X6:0i10}[530^m62c:g5c;326^0c4g3c5g431g6332c0g4~670e<540e=`31322}awm15_0g5c>‚G0c0g4e<g631322}awm15C0g6D6<0c?qg531530^5_/2e@cAqg3322cBg441;" #(#fn(io.write)
+	%fl:encode-byte-code #87=#fn("8000r1c0e1|3141;" #(#fn("8000r1c0e1|3141;" #(#fn(";000r1c0qe1c2|31b3c2|31b2VT2wc33241;" #(#fn("=000r1c0qc1~31`c230c230c330^^47;" #(#fn("?000r7c0g4c1322c2}|X6ÿ02i10}[m52g5c3‚O0c4g2i10}aw[c5g431332}b2wm15Ï0c0g4e6c7e8~6<0c9qg531540g53231322}awm12}|X6:0i10}[530^m62c:g5c;326^0c4g3c5g431g6332c0g4~670e<540e=`31322}awm15_0g5c>‚G0c0g4e<g631322}awm15C0g6D6<0c?qg531530^5_/2e@cAqg3322cBg441;" #(#fn(io.write)
   #int32(0) #{#<unspecified>}# label #fn(put!)
   #fn(sizeof) %fl:byte #fn(get) Instructions #fn("7000r1|c0‚50c1;|c2‚50c3;|c4‚50c5;|c6‚50c7;|c8‚50c9;|c:‚50c;;i05;" #(jmp
   jmp.l brt brt.l brf brf.l brne brne.l brnn brnn.l brn brn.l))
@@ -202,11 +207,11 @@
   #fn(io.tostring!))) #fn(length) #fn(table)
   #fn(buffer))) %fl:>= #fn(length) 65536)) %fl:list->vector)) %fl:reverse!) encode-byte-code)
 	%fl:enum #fn(enum) %fl:error
-	#86=#fn(":000s0c0c1|K41;" #(#fn(raise) error) error) %fl:eval #87=#fn("8000r1e0e1|313140;" #(%fl:compile-thunk
+	#88=#fn(":000s0c0c1|K41;" #(#fn(raise) error) error) %fl:eval #89=#fn("8000r1e0e1|313140;" #(%fl:compile-thunk
   %fl:expand) eval)
-	%fl:even? #88=#fn("8000r1c0|a32`W;" #(#fn(logand)) even?) %fl:every
-	#89=#fn("8000r2}?17D02|}M3116:02e0|}N42;" #(%fl:every) every)
-	%fl:expand #90=#fn("A000r1c0qc1c1c1c1c1c1c1c1c1c1c14;;" #(#fn("8000r;c0m02c1qm12c2L1m22c3qm32c4qm42c5qm52c6qm62c7qm72c8qm82c9m92c:qm:2g:~_42;" #(#fn("8000r2e0|31E17902c1|}32@;" #(resolve-global
+	%fl:even? #90=#fn("8000r1c0|a32`W;" #(#fn(logand)) even?) %fl:every
+	#91=#fn("8000r2}?17D02|}M3116:02e0|}N42;" #(%fl:every) every)
+	%fl:expand #92=#fn("A000r1c0qc1c1c1c1c1c1c1c1c1c1c14;;" #(#fn("8000r;c0m02c1qm12c2L1m22c3qm32c4qm42c5qm52c6qm62c7qm72c8qm82c9m92c:qm:2g:~_42;" #(#fn("8000r2e0|31E17902c1|}32@;" #(resolve-global
   #fn(assq)) top?) #fn("9000r1|?640|;|c0>640|;|MF16;02e1|31c2<6D0c3\x7fe4|3131\x7f|N3142;|M\x7f|N31K;" #(((begin))
   %fl:caar begin #fn(append) %fl:cdar) splice-begin) *expanded* #fn("9000r2|?640|;c0q~c1}32690\x7f|31530|41;" #(#fn("9000r1c0qi10c1\x7f3241;" #(#fn("8000r1c0q|6:0e1~31530_41;" #(#fn(":000r1c0qc1c2c3|32i213241;" #(#fn("8000r1i107=0c0c1qi2042;c2qc3qc431i203141;" #(#fn(map)
   #fn("8000r1i5:|~42;" #()) #fn("7000r1c0q|41;" #(#fn("9000r1c0|F6]02i62e1|31<7A0|i6:|Mi1032O590|e2|31O2|Nm05\x02/2~;" #(#{#<unspecified>}#
@@ -235,89 +240,89 @@
 			      #{#<unspecified>}#))))
 							      #fn(assq)))) expand-in)))
 								  #{#<unspecified>}#) expand)
-	%fl:expand-define #91=#fn("=000r1c0e1|31e2|31F6:0e2|315O0e1|31C6;0e330L15=0e4c5e6|313242;" #(#fn("<000r2|C6:0c0|}ML3;c0|Mc1c2L1|NL1c3}31|M34L3;" #(set!
+	%fl:expand-define #93=#fn("=000r1c0e1|31e2|31F6:0e2|315O0e1|31C6;0e330L15=0e4c5e6|313242;" #(#fn("<000r2|C6:0c0|}ML3;c0|Mc1c2L1|NL1c3}31|M34L3;" #(set!
   #fn(nconc) lambda #fn(copy-list))) %fl:cadr %fl:cddr %fl:void %fl:error "compile error: invalid syntax "
   %fl:print-to-string) expand-define)
-	%fl:expand-macro-call #92=#fn("7000r2e0690c1qc2t;|}Nv2;" #(*defer-macro-errors*
+	%fl:expand-macro-call #94=#fn("7000r2e0690c1qc2t;|}Nv2;" #(*defer-macro-errors*
 								   #fn("7000r0~\x7fNv2;" #())
 								   #fn("8000r1c0c1|L2L2;" #(raise
   quote))) expand-macro-call)
-	%fl:filter #93=#fn("7000r2c0qc141;" #(#fn("9000r1c0qm02|~\x7f_L143;" #(#fn("9000r3g2c0}F6T02i10}M316?0g2}M_KPNm2540c02}Nm15\x0b/2N;" #(#{#<unspecified>}#) filter-)))
+	%fl:filter #95=#fn("7000r2c0qc141;" #(#fn("9000r1c0qm02|~\x7f_L143;" #(#fn("9000r3g2c0}F6T02i10}M316?0g2}M_KPNm2540c02}Nm15\x0b/2N;" #(#{#<unspecified>}#) filter-)))
 					      #{#<unspecified>}#) filter)
-	%fl:fits-i8 #94=#fn("8000r1|I16F02e0|b°3216:02e1|b¯42;" #(%fl:>= %fl:<=) fits-i8)
+	%fl:fits-i8 #96=#fn("8000r1|I16F02e0|b°3216:02e1|b¯42;" #(%fl:>= %fl:<=) fits-i8)
 	%fl:float #fn(float) %fl:foldl
-	#95=#fn(":000r3g2…40};e0||g2M}32g2N43;" #(%fl:foldl) foldl) %fl:foldr
-	#96=#fn(";000r3g2…40};|g2Me0|}g2N3342;" #(%fl:foldr) foldr)
-	%fl:for-each #97=#fn(";000s2c0qc141;" #(#fn(":000r1c0qm02i02…K0c1\x7fF6A02~\x7fM312\x7fNo015\x1e/5;0|~\x7fi02K322];" #(#fn(":000r2}MF6I0|c0c1}32Q22~|c0c2}3242;c3;" #(#fn(map)
+	#97=#fn(":000r3g2…40};e0||g2M}32g2N43;" #(%fl:foldl) foldl) %fl:foldr
+	#98=#fn(";000r3g2…40};|g2Me0|}g2N3342;" #(%fl:foldr) foldr)
+	%fl:for-each #99=#fn(";000s2c0qc141;" #(#fn(":000r1c0qm02i02…K0c1\x7fF6A02~\x7fM312\x7fNo015\x1e/5;0|~\x7fi02K322];" #(#fn(":000r2}MF6I0|c0c1}32Q22~|c0c2}3242;c3;" #(#fn(map)
   #.car #.cdr #{#<unspecified>}#) for-each-n) #{#<unspecified>}#))
 						#{#<unspecified>}#) for-each)
-	%fl:function-source #98=#fn("8000r1|J16D02|G@16<02c0c1|3141;" #(#fn("9000r1e0c1|31`3216@02c2|c1|31ax[41;" #(%fl:>
+	%fl:function-source #100=#fn("8000r1|J16D02|G@16<02c0c1|3141;" #(#fn("9000r1e0c1|31`3216@02c2|c1|31ax[41;" #(%fl:>
   #fn(length) #fn("7000r1|F16?02|Mc0<16502|N;" #(%source))))
   #fn(function:vals)) function-source)
-	%fl:get-defined-vars #99=#fn("8000r1e0~|3141;" #(%fl:delete-duplicates) #1=#(#fn("9000r1|?640_;|Mc0<16602|NF6u0e1|31C16:02e1|31L117^02e1|31F16M02e2e1|3131C16>02e2e1|3131L117402_;|Mc3‚>0c4c5~|N32v2;_;" #(define
+	%fl:get-defined-vars #101=#fn("8000r1e0~|3141;" #(%fl:delete-duplicates) #1=#(#fn("9000r1|?640_;|Mc0<16602|NF6u0e1|31C16:02e1|31L117^02e1|31F16M02e2e1|3131C16>02e2e1|3131L117402_;|Mc3‚>0c4c5~|N32v2;_;" #(define
   %fl:cadr %fl:define-head-name begin #fn(nconc)
   #fn(map)) #1#) ()))
-	%fl:hex5 #100=#fn("9000r1e0c1|b@32b5c243;" #(%fl:string.lpad #fn(number->string)
+	%fl:hex5 #102=#fn("9000r1e0c1|b@32b5c243;" #(%fl:string.lpad #fn(number->string)
 						     #\0) hex5)
-	%fl:identity #101=#fn("6000r1|;" #() identity) %fl:in-env?
-	#102=#fn("8000r2}F16F02c0|}M3217:02e1|}N42;" #(#fn(memq) %fl:in-env?) in-env?)
-	%fl:index-of #103=#fn(":000r3}…40^;|}M‚50g2;e0|}Ng2aw43;" #(%fl:index-of) index-of)
+	%fl:identity #103=#fn("6000r1|;" #() identity) %fl:in-env?
+	#104=#fn("8000r2}F16F02c0|}M3217:02e1|}N42;" #(#fn(memq) %fl:in-env?) in-env?)
+	%fl:index-of #105=#fn(":000r3}…40^;|}M‚50g2;e0|}Ng2aw43;" #(%fl:index-of) index-of)
 	%fl:int16 #fn(int16) %fl:int32
 	#fn(int32) %fl:int64 #fn(int64) %fl:int8
-	#fn(int8) %fl:io.readall #104=#fn("7000r1c0qc13041;" #(#fn("8000r1c0|~322c1qc2|3141;" #(#fn(io.copy)
+	#fn(int8) %fl:io.readall #106=#fn("7000r1c0qc13041;" #(#fn("8000r1c0|~322c1qc2|3141;" #(#fn(io.copy)
   #fn("7000r1|c0>16:02c1i1031670c240;|;" #("" #fn(io.eof?)
 					   #fn(eof-object)))
   #fn(io.tostring!))) #fn(buffer)) io.readall)
-	%fl:io.readline #105=#fn("8000r1c0|c142;" #(#fn(io.readuntil) #\newline) io.readline)
-	%fl:io.readlines #106=#fn("8000r1e0e1|42;" #(%fl:read-all-of
+	%fl:io.readline #107=#fn("8000r1c0|c142;" #(#fn(io.readuntil) #\newline) io.readline)
+	%fl:io.readlines #108=#fn("8000r1e0e1|42;" #(%fl:read-all-of
 						     %fl:io.readline) io.readlines)
-	%fl:iota #107=#fn("8000r1e0e1|42;" #(%fl:map-int %fl:identity) iota)
-	%fl:keyword->symbol #108=#fn("9000r1c0|316@0c1c2c3|313141;|;" #(#fn(keyword?)
+	%fl:iota #109=#fn("8000r1e0e1|42;" #(%fl:map-int %fl:identity) iota)
+	%fl:keyword->symbol #110=#fn("9000r1c0|316@0c1c2c3|313141;|;" #(#fn(keyword?)
   #fn(symbol) #fn("<000r1c0|`c1|c2|313243;" #(#fn(string.sub)
 					      #fn(string.dec)
 					      #fn(length)))
   #fn(string)) keyword->symbol)
-	%fl:keyword-arg? #109=#fn("7000r1|F16902c0|M41;" #(#fn(keyword?)) keyword-arg?)
-	%fl:lambda-arg-names #110=#fn("9000r1e0c1e2|3142;" #(%fl:map! #fn("7000r1|F690e0|M41;|;" #(%fl:keyword->symbol))
+	%fl:keyword-arg? #111=#fn("7000r1|F16902c0|M41;" #(#fn(keyword?)) keyword-arg?)
+	%fl:lambda-arg-names #112=#fn("9000r1e0c1e2|3142;" #(%fl:map! #fn("7000r1|F690e0|M41;|;" #(%fl:keyword->symbol))
 							     %fl:to-proper) lambda-arg-names)
-	%fl:lambda-vars #111=#fn("7000r1c0qc141;" #(#fn(":000r1c0qm02|~~^^342e1~41;" #(#fn(";000r4|A17502|C640];|F16602|MC6S0g217502g36<0e0c1}c243;~|N}g2g344;|F16602|MF6‡0e3|Mb23216902e4|31C660^5=0e0c5|Mc6}342c7e4|31316<0~|N}g2]44;g36<0e0c1}c843;~|N}]g344;|F6>0e0c9|Mc6}44;|}‚:0e0c1}42;e0c9|c6}44;" #(%fl:error
+	%fl:lambda-vars #113=#fn("7000r1c0qc141;" #(#fn(":000r1c0qm02|~~^^342e1~41;" #(#fn(";000r4|A17502|C640];|F16602|MC6S0g217502g36<0e0c1}c243;~|N}g2g344;|F16602|MF6‡0e3|Mb23216902e4|31C660^5=0e0c5|Mc6}342c7e4|31316<0~|N}g2]44;g36<0e0c1}c843;~|N}]g344;|F6>0e0c9|Mc6}44;|}‚:0e0c1}42;e0c9|c6}44;" #(%fl:error
   "compile error: invalid argument list "
   ". optional arguments must come after required." %fl:length= %fl:caar "compile error: invalid optional argument "
   " in list " #fn(keyword?) ". keyword arguments must come last."
   "compile error: invalid formal argument ") check-formals)
   %fl:lambda-arg-names)) #{#<unspecified>}#) lambda-vars)
-	%fl:last-pair #112=#fn("7000r1|N?640|;e0|N41;" #(%fl:last-pair) last-pair)
-	%fl:lastcdr #113=#fn("7000r1|?640|;e0|31N;" #(%fl:last-pair) lastcdr)
-	%fl:length= #114=#fn("9000r2}`X640^;}`W650|?;|?660}`W;e0|N}ax42;" #(%fl:length=) length=)
-	%fl:length> #115=#fn("9000r2}`X640|;}`W6;0|F16402|;|?660}`X;e0|N}ax42;" #(%fl:length>) length>)
-	%fl:list->vector #116=#fn("7000r1c0|v2;" #(#.vector) list->vector)
-	%fl:list-head #117=#fn(":000r2e0}`32640_;|Me1|N}ax32K;" #(%fl:<=
+	%fl:last-pair #114=#fn("7000r1|N?640|;e0|N41;" #(%fl:last-pair) last-pair)
+	%fl:lastcdr #115=#fn("7000r1|?640|;e0|31N;" #(%fl:last-pair) lastcdr)
+	%fl:length= #116=#fn("9000r2}`X640^;}`W650|?;|?660}`W;e0|N}ax42;" #(%fl:length=) length=)
+	%fl:length> #117=#fn("9000r2}`X640|;}`W6;0|F16402|;|?660}`X;e0|N}ax42;" #(%fl:length>) length>)
+	%fl:list->vector #118=#fn("7000r1c0|v2;" #(#.vector) list->vector)
+	%fl:list-head #119=#fn(":000r2e0}`32640_;|Me1|N}ax32K;" #(%fl:<=
 								  %fl:list-head) list-head)
-	%fl:list-ref #118=#fn("8000r2e0|}32M;" #(%fl:list-tail) list-ref)
-	%fl:list-tail #119=#fn("9000r2e0}`32640|;e1|N}ax42;" #(%fl:<=
+	%fl:list-ref #120=#fn("8000r2e0|}32M;" #(%fl:list-tail) list-ref)
+	%fl:list-tail #121=#fn("9000r2e0}`32640|;e1|N}ax42;" #(%fl:<=
 							       %fl:list-tail) list-tail)
-	%fl:list? #120=#fn("7000r1|A17@02|F16902e0|N41;" #(%fl:list?) list?)
-	%fl:load #121=#fn("9000r1c0qc1|c23241;" #(#fn("7000r1c0qc1qt;" #(#fn("9000r0c0qc131c1c1c143;" #(#fn("6000r1c0qm02|;" #(#fn(":000r3c0i10317C0~c1i1031|e2}3143;c3i10312e2}41;" #(#fn(io.eof?)
+	%fl:list? #122=#fn("7000r1|A17@02|F16902e0|N41;" #(%fl:list?) list?)
+	%fl:load #123=#fn("9000r1c0qc1|c23241;" #(#fn("7000r1c0qc1qt;" #(#fn("9000r0c0qc131c1c1c143;" #(#fn("6000r1c0qm02|;" #(#fn(":000r3c0i10317C0~c1i1031|e2}3143;c3i10312e2}41;" #(#fn(io.eof?)
   #fn(read) %fl:load-process #fn(io.close))))) #{#<unspecified>}#))
   #fn("9000r1c0~312c1c2i10|L341;" #(#fn(io.close)
 				    #fn(raise) load-error))))
 						  #fn(file) :read) load)
-	%fl:load-process #122=#fn("7000r1e0|41;" #(%fl:eval) load-process)
+	%fl:load-process #124=#fn("7000r1e0|41;" #(%fl:eval) load-process)
 	%fl:long #fn(long) %fl:lookup-sym
-	#123=#fn("7000r4}…50c0;c1q}M41;" #((global)
+	#125=#fn("7000r4}…50c0;c1q}M41;" #((global)
 					   #fn(":000r1c0qe1~|`3341;" #(#fn(";000r1|6@0i13640|;i12|K;e0i10i11Ni1317502~A680i12570i12aw^44;" #(%fl:lookup-sym))
   %fl:index-of))) lookup-sym)
-	%fl:macrocall? #124=#fn("8000r1|MC16=02e0e1|M3141;" #(%fl:symbol-syntax
+	%fl:macrocall? #126=#fn("8000r1|MC16=02e0e1|M3141;" #(%fl:symbol-syntax
 							      resolve-global) macrocall?)
-	%fl:macroexpand-1 #125=#fn("8000r1|?640|;c0qe1|3141;" #(#fn("7000r1|680|~Nv2;~;" #())
+	%fl:macroexpand-1 #127=#fn("8000r1|?640|;c0qe1|3141;" #(#fn("7000r1|680|~Nv2;~;" #())
 								%fl:macrocall?) macroexpand-1)
-	%fl:make-code-emitter #126=#fn("9000r0_c030`c1Z4;" #(#fn(table) +inf.0) make-code-emitter)
-	%fl:make-label #127=#fn("6000r1c040;" #(#fn(gensym)) make-label)
-	%fl:make-perfect-hash-table #128=#fn("7000r1c0qc141;" #(#fn("8000r1c0m02c1qc231c3~3141;" #(#fn("9000r2e0e1c2|3131}42;" #(%fl:mod0
+	%fl:make-code-emitter #128=#fn("9000r0_c030`c1Z4;" #(#fn(table) +inf.0) make-code-emitter)
+	%fl:make-label #129=#fn("6000r1c040;" #(#fn(gensym)) make-label)
+	%fl:make-perfect-hash-table #130=#fn("7000r1c0qc141;" #(#fn("8000r1c0m02c1qc231c3~3141;" #(#fn("9000r2e0e1c2|3131}42;" #(%fl:mod0
   %fl:abs #fn(hash)) $hash-keyword) #fn("6000r1c0qm02|;" #(#fn("9000r1c0qc1b2|T2^3241;" #(#fn("7000r1c0qc131i3041;" #(#fn("6000r1c0qm02|;" #(#fn("8000r1|F6=0c0qe1|3141;i10;" #(#fn(":000r1c0qb2i50|i3032T241;" #(#fn("9000r1i30|[6=0i50i40aw41;i30|~\\2i30|awe0i1031\\2i20i10N41;" #(%fl:cdar))))
   %fl:caar)))) #{#<unspecified>}#)) #fn(vector.alloc))))) #{#<unspecified>}# #fn(length)))
 								#{#<unspecified>}#) make-perfect-hash-table)
-	%fl:make-system-image #129=#fn(";000r1c0c1|c2c3c434c542;" #(#fn("8000r2c0qe1e242;" #(#fn("7000r2]k02]k12c2qc3q41;" #(*print-pretty*
+	%fl:make-system-image #131=#fn(";000r1c0c1|c2c3c434c542;" #(#fn("8000r2c0qe1e242;" #(#fn("7000r2]k02]k12c2qc3q41;" #(*print-pretty*
   *print-readably* #fn("7000r1c0qc1qt|302;" #(#fn(":000r0c0qe1c2qe3c4303132312c5i2041;" #(#fn("=000r1c0c1c2c3|c2c4|3233Q2i20322c5i20e642;" #(#fn(write)
   #fn(nconc) #fn(map) #.list #fn(top-level-value)
   #fn(io.write) *linefeed*)) %fl:filter #fn("9000r1|E16w02c0|31@16l02c1|31G@17C02c2|31c2c1|3131>@16K02c3|i2132@16=02c4c1|3131@;" #(#fn(constant?)
@@ -334,39 +339,39 @@
 							 *print-level*
 							 *print-length*
 							 *os-name*)) make-system-image)
-	%fl:map! #130=#fn("9000r2}c0}F6B02}|}M31O2}Nm15\x1d/2;" #(#{#<unspecified>}#) map!)
-	%fl:map-int #131=#fn("8000r2e0}`32640_;c1q|`31_K_42;" #(%fl:<= #fn(":000r2|m12a\x7faxc0qu2|;" #(#fn("8000r1\x7fi10|31_KP2\x7fNo01;" #())))) map-int)
-	%fl:mark-label #132=#fn("9000r2e0|c1}43;" #(%fl:emit label) mark-label)
-	%fl:max #133=#fn("<000s1}…40|;e0c1|}43;" #(%fl:foldl #fn("7000r2|}X640};|;" #())) max)
-	%fl:member #134=#fn("8000r2}?640^;}M|>640};e0|}N42;" #(%fl:member) member)
-	%fl:memv #135=#fn("8000r2}?640^;}M|=640};e0|}N42;" #(%fl:memv) memv)
-	%fl:min #136=#fn("<000s1}…40|;e0c1|}43;" #(%fl:foldl #fn("7000r2|}X640|;};" #())) min)
-	%fl:mod #137=#fn("9000r2|e0|}32}T2x;" #(%fl:div) mod) %fl:mod0
-	#138=#fn("8000r2||}V}T2x;" #() mod0) %fl:nan? #139=#fn("7000r1|c0>17702|c1>;" #(+nan.0
+	%fl:map! #132=#fn("9000r2}c0}F6B02}|}M31O2}Nm15\x1d/2;" #(#{#<unspecified>}#) map!)
+	%fl:map-int #133=#fn("8000r2e0}`32640_;c1q|`31_K_42;" #(%fl:<= #fn(":000r2|m12a\x7faxc0qu2|;" #(#fn("8000r1\x7fi10|31_KP2\x7fNo01;" #())))) map-int)
+	%fl:mark-label #134=#fn("9000r2e0|c1}43;" #(%fl:emit label) mark-label)
+	%fl:max #135=#fn("<000s1}…40|;e0c1|}43;" #(%fl:foldl #fn("7000r2|}X640};|;" #())) max)
+	%fl:member #136=#fn("8000r2}?640^;}M|>640};e0|}N42;" #(%fl:member) member)
+	%fl:memv #137=#fn("8000r2}?640^;}M|=640};e0|}N42;" #(%fl:memv) memv)
+	%fl:min #138=#fn("<000s1}…40|;e0c1|}43;" #(%fl:foldl #fn("7000r2|}X640|;};" #())) min)
+	%fl:mod #139=#fn("9000r2|e0|}32}T2x;" #(%fl:div) mod) %fl:mod0
+	#140=#fn("8000r2||}V}T2x;" #() mod0) %fl:nan? #141=#fn("7000r1|c0>17702|c1>;" #(+nan.0
   -nan.0) nan?)
-	%fl:nary-compare #140=#fn("9000r2}A17Q02}NA17I02|}Me0}313216:02e1|}N42;" #(%fl:cadr
+	%fl:nary-compare #142=#fn("9000r2}A17Q02}NA17I02|}Me0}313216:02e1|}N42;" #(%fl:cadr
   %fl:nary-compare) nary-compare)
-	%fl:nary< #141=#fn(":000s0e0c1|42;" #(%fl:nary-compare #fn("7000r2|}X;" #())) nary<)
-	%fl:nary= #142=#fn(":000s0e0c1|42;" #(%fl:nary-compare #fn("7000r2|}W;" #())) nary=)
-	%fl:negative? #143=#fn("7000r1|`X;" #() negative?) %fl:nestlist
-	#144=#fn(";000r3e0g2`32640_;}e1||}31g2ax33K;" #(%fl:<= %fl:nestlist) nestlist)
-	%fl:newline #145=#fn("9000‰00001000Š0000770e0m02c1|e2322];" #(*output-stream*
+	%fl:nary< #143=#fn(":000s0e0c1|42;" #(%fl:nary-compare #fn("7000r2|}X;" #())) nary<)
+	%fl:nary= #144=#fn(":000s0e0c1|42;" #(%fl:nary-compare #fn("7000r2|}W;" #())) nary=)
+	%fl:negative? #145=#fn("7000r1|`X;" #() negative?) %fl:nestlist
+	#146=#fn(";000r3e0g2`32640_;}e1||}31g2ax33K;" #(%fl:<= %fl:nestlist) nestlist)
+	%fl:newline #147=#fn("9000‰00001000Š0000770e0m02c1|e2322];" #(*output-stream*
   #fn(io.write) *linefeed*) newline)
-	%fl:nnn #146=#fn("8000r1e0c1|42;" #(%fl:count #fn("6000r1|A@;" #())) nnn)
-	%fl:nreconc #147=#fn("8000r2e0}|42;" #(%fl:reverse!-) nreconc) %fl:odd?
-	#148=#fn("7000r1e0|31@;" #(%fl:even?) odd?) %fl:positive? #149=#fn("8000r1e0|`42;" #(%fl:>) positive?)
-	%fl:princ #150=#fn("9000s0c0qe141;" #(#fn("7000r1^k02c1qc2q41;" #(*print-readably*
+	%fl:nnn #148=#fn("8000r1e0c1|42;" #(%fl:count #fn("6000r1|A@;" #())) nnn)
+	%fl:nreconc #149=#fn("8000r2e0}|42;" #(%fl:reverse!-) nreconc) %fl:odd?
+	#150=#fn("7000r1e0|31@;" #(%fl:even?) odd?) %fl:positive? #151=#fn("8000r1e0|`42;" #(%fl:>) positive?)
+	%fl:princ #152=#fn("9000s0c0qe141;" #(#fn("7000r1^k02c1qc2q41;" #(*print-readably*
   #fn("7000r1c0qc1qt|302;" #(#fn("8000r0e0c1i2042;" #(%fl:for-each #fn(write)))
 			     #fn("7000r1~302c0|41;" #(#fn(raise)))))
   #fn("6000r0~k0;" #(*print-readably*)))) *print-readably*) princ)
-	%fl:print #151=#fn(":000s0e0c1|42;" #(%fl:for-each #fn(write)) print)
-	%fl:print-exception #152=#fn("=000r1|F16D02|Mc0<16:02e1|b4326S0e2c3e4|31c5e6|31c7352e8e9|31315\x130|F16D02|Mc:<16:02e1|b4326Q0e2e4|31c;e9|31c<342e8e6|31315Ù0|F16@02|Mc=<16602|NF6B0e2c>e4|31c?335²0|F16802|Mc@<6B0e2cA312e2|NQ25“0|F16802|McB<6J0eCe6|31312e2cDe4|31325l0eE|3116:02e1|b2326L0e8|M312e2cF312cGe4|31315>0e2cH312e8|312e2eI41;" #(type-error
+	%fl:print #153=#fn(":000s0e0c1|42;" #(%fl:for-each #fn(write)) print)
+	%fl:print-exception #154=#fn("=000r1|F16D02|Mc0<16:02e1|b4326S0e2c3e4|31c5e6|31c7352e8e9|31315\x130|F16D02|Mc:<16:02e1|b4326Q0e2e4|31c;e9|31c<342e8e6|31315Ù0|F16@02|Mc=<16602|NF6B0e2c>e4|31c?335²0|F16802|Mc@<6B0e2cA312e2|NQ25“0|F16802|McB<6J0eCe6|31312e2cDe4|31325l0eE|3116:02e1|b2326L0e8|M312e2cF312cGe4|31315>0e2cH312e8|312e2eI41;" #(type-error
   %fl:length= %fl:princ "type error: " %fl:cadr ": expected " %fl:caddr ", got "
   %fl:print %fl:cadddr bounds-error ": index " " out of bounds for "
   unbound-error "eval: variable " " has no value" error "error: " load-error
   %fl:print-exception "in file " %fl:list? ": " #fn("8000r1c0|3117502|C670e1540e2|41;" #(#fn(string?)
   %fl:princ %fl:print)) "*** Unhandled exception: " *linefeed*) print-exception)
-	%fl:print-stack-trace #153=#fn("8000r1c0qc1c142;" #(#fn("=000r2c0qm02c1qm12c2qe3e4~e5670b5540b43231e6e7c8c9c:303232`43;" #(#fn("8000r3c0qc1|31g2K41;" #(#fn("9000r1c0~31c0\x7f31‚>0c1c2c3|L341;c4qc5~3141;" #(#fn(function:code)
+	%fl:print-stack-trace #155=#fn("8000r1c0qc1c142;" #(#fn("=000r2c0qm02c1qm12c2qe3e4~e5670b5540b43231e6e7c8c9c:303232`43;" #(#fn("8000r3c0qc1|31g2K41;" #(#fn("9000r1c0~31c0\x7f31‚>0c1c2c3|L341;c4qc5~3141;" #(#fn(function:code)
   #fn(raise) thrown-value ffound #fn(":000r1`e0c1|3131c2qu;" #(%fl:1- #fn(length)
 							       #fn("9000r1e0~|[316A0i30~|[i21i1043;c1;" #(%fl:closure?
   #{#<unspecified>}#)))) #fn(function:vals)))
@@ -380,24 +385,24 @@
   *interactive* %fl:filter %fl:closure? #fn(map)
   #fn("7000r1|E16802c0|41;" #(#fn(top-level-value)))
   #fn(environment))) #{#<unspecified>}#) print-stack-trace)
-	%fl:print-to-string #154=#fn("7000r1c0qc13041;" #(#fn("8000r1c0~|322c1|41;" #(#fn(write)
+	%fl:print-to-string #156=#fn("7000r1c0qc13041;" #(#fn("8000r1c0~|322c1|41;" #(#fn(write)
   #fn(io.tostring!))) #fn(buffer)) print-to-string)
-	%fl:printable? #155=#fn("7000r1c0|3117802c1|31@;" #(#fn(iostream?)
+	%fl:printable? #157=#fn("7000r1c0|3117802c1|31@;" #(#fn(iostream?)
 							    #fn(eof-object?)) printable?)
-	%fl:proper-part #156=#fn("8000r1|F6<0|Me0|N31K;_;" #(%fl:proper-part) proper-part)
-	%fl:quote-value #157=#fn("7000r1e0|31640|;c1|L2;" #(%fl:self-evaluating?
+	%fl:proper-part #158=#fn("8000r1|F6<0|Me0|N31K;_;" #(%fl:proper-part) proper-part)
+	%fl:quote-value #159=#fn("7000r1e0|31640|;c1|L2;" #(%fl:self-evaluating?
 							    quote) quote-value)
-	%fl:random #158=#fn("8000r1c0|316<0e1c230|42;c330|T2;" #(#fn(integer?)
+	%fl:random #160=#fn("8000r1c0|316<0e1c230|42;c330|T2;" #(#fn(integer?)
 								 %fl:mod #fn(rand)
 								 #fn(rand.double)) random)
-	%fl:read-all #159=#fn("8000r1e0c1|42;" #(%fl:read-all-of #fn(read)) read-all)
-	%fl:read-all-of #160=#fn("9000r2c0qc131_|}3142;" #(#fn("6000r1c0qm02|;" #(#fn("9000r2c0i1131680e1|41;~}|Ki10i113142;" #(#fn(io.eof?)
+	%fl:read-all #161=#fn("8000r1e0c1|42;" #(%fl:read-all-of #fn(read)) read-all)
+	%fl:read-all-of #162=#fn("9000r2c0qc131_|}3142;" #(#fn("6000r1c0qm02|;" #(#fn("9000r2c0i1131680e1|41;~}|Ki10i113142;" #(#fn(io.eof?)
   %fl:reverse!)))) #{#<unspecified>}#) read-all-of)
-	%fl:ref-int16-LE #161=#fn(";000r2e0c1|}`w[`32c1|}aw[b832w41;" #(%fl:int16
+	%fl:ref-int16-LE #163=#fn(";000r2e0c1|}`w[`32c1|}aw[b832w41;" #(%fl:int16
   #fn(ash)) ref-int16-LE)
-	%fl:ref-int32-LE #162=#fn("=000r2e0c1|}`w[`32c1|}aw[b832c1|}b2w[b@32c1|}b3w[bH32R441;" #(%fl:int32
+	%fl:ref-int32-LE #164=#fn("=000r2e0c1|}`w[`32c1|}aw[b832c1|}b2w[b@32c1|}b3w[bH32R441;" #(%fl:int32
   #fn(ash)) ref-int32-LE)
-	%fl:repl #163=#fn("8000r0c0c1c142;" #(#fn("6000r2c0m02c1qm12}302e240;" #(#fn("8000r0e0c1312c2e3312c4c5c6t41;" #(%fl:princ
+	%fl:repl #165=#fn("8000r0c0c1c142;" #(#fn("6000r2c0m02c1qm12}302e240;" #(#fn("8000r0e0c1312c2e3312c4c5c6t41;" #(%fl:princ
   "> " #fn(io.flush) *output-stream* #fn("8000r1c0e131@16<02c2e3|3141;" #(#fn(io.eof?)
   *input-stream* #fn("7000r1e0|312|k12];" #(%fl:print that)) %fl:load-process))
   #fn("6000r0c040;" #(#fn(read))) #fn("7000r1c0e1312c2|41;" #(#fn(io.discardbuffer)
@@ -406,87 +411,92 @@
 					#fn("7000r1e0|312];" #(%fl:top-level-exception-handler))
 					%fl:newline) reploop) %fl:newline))
 					      #{#<unspecified>}#) repl)
-	%fl:revappend #164=#fn("8000r2e0}|42;" #(%fl:reverse-) revappend)
-	%fl:reverse #165=#fn("8000r1e0_|42;" #(%fl:reverse-) reverse)
-	%fl:reverse! #166=#fn("8000r1e0_|42;" #(%fl:reverse!-) reverse!)
-	%fl:reverse!- #167=#fn("9000r2c0}F6B02}N}|}m02P2m15\x1d/2|;" #(#{#<unspecified>}#) reverse!-)
-	%fl:reverse- #168=#fn("8000r2}…40|;e0}M|K}N42;" #(%fl:reverse-) reverse-)
-	%fl:self-evaluating? #169=#fn("8000r1|?16602|C@17K02c0|3116A02|C16:02|c1|31<;" #(#fn(constant?)
+	%fl:revappend #166=#fn("8000r2e0}|42;" #(%fl:reverse-) revappend)
+	%fl:reverse #167=#fn("8000r1e0_|42;" #(%fl:reverse-) reverse)
+	%fl:reverse! #168=#fn("8000r1e0_|42;" #(%fl:reverse!-) reverse!)
+	%fl:reverse!- #169=#fn("9000r2c0}F6B02}N}|}m02P2m15\x1d/2|;" #(#{#<unspecified>}#) reverse!-)
+	%fl:reverse- #170=#fn("8000r2}…40|;e0}M|K}N42;" #(%fl:reverse-) reverse-)
+	%fl:self-evaluating? #171=#fn("8000r1|?16602|C@17K02c0|3116A02|C16:02|c1|31<;" #(#fn(constant?)
   #fn(top-level-value)) self-evaluating?)
-	%fl:separate #170=#fn("7000r2c0qc141;" #(#fn(":000r1c0m02|~\x7f_L1_L144;" #(#fn(";000r4c0g2g3Kc1}F6Z02|}M316?0g2}M_KPNm25<0g3}M_KPNm32}Nm15\x05/241;" #(#fn("8000r1e0|MN|NN42;" #(%fl:values))
+	%fl:separate #172=#fn("7000r2c0qc141;" #(#fn(":000r1c0m02|~\x7f_L1_L144;" #(#fn(";000r4c0g2g3Kc1}F6Z02|}M316?0g2}M_KPNm25<0g3}M_KPNm32}Nm15\x05/241;" #(#fn("8000r1e0|MN|NN42;" #(%fl:values))
   #{#<unspecified>}#) separate-))) #{#<unspecified>}#) separate)
-	%fl:set-syntax! #171=#fn("9000r2c0e1e2|31}43;" #(#fn(put!)
+	%fl:set-syntax! #173=#fn("9000r2c0e1e2|31}43;" #(#fn(put!)
 							 *syntax-environment*
 							 resolve-global) set-syntax!)
-	%fl:simple-sort #172=#fn("7000r1|A17602|NA640|;c0q|M41;" #(#fn("8000r1e0c1qc2q42;" #(%fl:call-with-values
+	%fl:simple-compare #174=#fn("7000r2c0q|M41;" #(#fn(":000r1c0|c13216x02e2~b33216l02e3|\x7f32@16`02e4|31|<16T02c5qc631c73116D02c8qc9e:~31e;~3143;" #(#fn(memq)
+  (> <= >=) %fl:length= %fl:in-env? resolve-global #fn("6000r1c0qm02|;" #(#fn("8000r1|A17U02e0|Mi2132@16F02e1|M31|M<16802~|N41;" #(%fl:in-env?
+  resolve-global)))) #{#<unspecified>}# (< = number?)
+  #fn("7000r3|}3116C02|g23116902c0q~41;" #(#fn("8000r1|c0=6;0c1i02\x7fL3;|c2=6;0e3\x7fi0242;e3i02\x7f42;" #(>
+  < <= %fl:compare<=)))) #fn("7000r1|?17802|Mc0<;" #(quote)) %fl:cadr %fl:caddr))) simple-compare)
+	%fl:simple-sort #175=#fn("7000r1|A17602|NA640|;c0q|M41;" #(#fn("8000r1e0c1qc2q42;" #(%fl:call-with-values
   #fn("8000r0e0c1qi10N42;" #(%fl:separate #fn("7000r1|~X;" #())))
   #fn(":000r2c0e1|31~L1e1}3143;" #(#fn(nconc) %fl:simple-sort))))) simple-sort)
-	%fl:splice-form? #173=#fn("8000r1|F16X02|Mc0<17N02|Mc1<17D02|Mc2<16:02e3|b23217702|c2<;" #(unquote-splicing
+	%fl:splice-form? #176=#fn("8000r1|F16X02|Mc0<17N02|Mc1<17D02|Mc2<16:02e3|b23217702|c2<;" #(unquote-splicing
   unquote-nsplicing unquote %fl:length>) splice-form?)
-	%fl:string.join #174=#fn("7000r2|…50c0;c1qc23041;" #("" #fn("8000r1c0|~M322e1c2q~N322c3|41;" #(#fn(io.write)
+	%fl:string.join #177=#fn("7000r2|…50c0;c1qc23041;" #("" #fn("8000r1c0|~M322e1c2q~N322c3|41;" #(#fn(io.write)
   %fl:for-each #fn("8000r1c0~i11322c0~|42;" #(#fn(io.write)))
   #fn(io.tostring!))) #fn(buffer)) string.join)
-	%fl:string.lpad #175=#fn(";000r3c0e1g2}c2|31x32|42;" #(#fn(string)
+	%fl:string.lpad #178=#fn(";000r3c0e1g2}c2|31x32|42;" #(#fn(string)
 							       %fl:string.rep
 							       #fn(string.count)) string.lpad)
-	%fl:string.map #176=#fn("9000r2c0qc130c2}3142;" #(#fn("7000r2c0q`312c1|41;" #(#fn(";000r1c0|\x7fX6S02c1~i10c2i11|3231322c3i11|32m05\x0b/;" #(#{#<unspecified>}#
+	%fl:string.map #179=#fn("9000r2c0qc130c2}3142;" #(#fn("7000r2c0q`312c1|41;" #(#fn(";000r1c0|\x7fX6S02c1~i10c2i11|3231322c3i11|32m05\x0b/;" #(#{#<unspecified>}#
   #fn(io.putc) #fn(string.char) #fn(string.inc)))
   #fn(io.tostring!))) #fn(buffer) #fn(length)) string.map)
-	%fl:string.rep #177=#fn(";000r2}b4X6`0e0}`32650c1;}aW680c2|41;}b2W690c2||42;c2|||43;e3}316@0c2|e4|}ax3242;e4c2||32}b2U242;" #(%fl:<=
+	%fl:string.rep #180=#fn(";000r2}b4X6`0e0}`32650c1;}aW680c2|41;}b2W690c2||42;c2|||43;e3}316@0c2|e4|}ax3242;e4c2||32}b2U242;" #(%fl:<=
   "" #fn(string) %fl:odd? %fl:string.rep) string.rep)
-	%fl:string.rpad #178=#fn("<000r3c0|e1g2}c2|31x3242;" #(#fn(string)
+	%fl:string.rpad #181=#fn("<000r3c0|e1g2}c2|31x3242;" #(#fn(string)
 							       %fl:string.rep
 							       #fn(string.count)) string.rpad)
-	%fl:string.tail #179=#fn(";000r2c0|c1|`}3342;" #(#fn(string.sub)
+	%fl:string.tail #182=#fn(";000r2c0|c1|`}3342;" #(#fn(string.sub)
 							 #fn(string.inc)) string.tail)
-	%fl:string.trim #180=#fn("8000r3c0qc1c142;" #(#fn("8000r2c0qm02c1qm12c2qc3~3141;" #(#fn(";000r4g2g3X16?02c0}c1|g232326A0~|}c2|g232g344;g2;" #(#fn(string.find)
+	%fl:string.trim #183=#fn("8000r3c0qc1c142;" #(#fn("8000r2c0qm02c1qm12c2qc3~3141;" #(#fn(";000r4g2g3X16?02c0}c1|g232326A0~|}c2|g232g344;g2;" #(#fn(string.find)
   #fn(string.char) #fn(string.inc)) trim-start)
   #fn("<000r3e0g2`3216D02c1}c2|c3|g23232326?0\x7f|}c3|g23243;g2;" #(%fl:> #fn(string.find)
 								    #fn(string.char)
 								    #fn(string.dec)) trim-end)
   #fn("<000r1c0i10~i10i11`|34\x7fi10i12|3343;" #(#fn(string.sub)))
   #fn(length))) #{#<unspecified>}#) string.trim)
-	%fl:symbol-syntax #181=#fn("9000r1c0e1|^43;" #(#fn(get)
+	%fl:symbol-syntax #184=#fn("9000r1c0e1|^43;" #(#fn(get)
 						       *syntax-environment*) symbol-syntax)
-	%fl:table.clone #182=#fn("7000r1c0qc13041;" #(#fn("9000r1c0c1q_~332|;" #(#fn(table.foldl)
+	%fl:table.clone #185=#fn("7000r1c0qc13041;" #(#fn("9000r1c0c1q_~332|;" #(#fn(table.foldl)
   #fn("9000r3c0~|}43;" #(#fn(put!))))) #fn(table)) table.clone)
-	%fl:table.foreach #183=#fn("9000r2c0c1q_}43;" #(#fn(table.foldl)
+	%fl:table.foreach #186=#fn("9000r2c0c1q_}43;" #(#fn(table.foldl)
 							#fn("8000r3~|}322];" #())) table.foreach)
-	%fl:table.invert #184=#fn("7000r1c0qc13041;" #(#fn("9000r1c0c1q_~332|;" #(#fn(table.foldl)
+	%fl:table.invert #187=#fn("7000r1c0qc13041;" #(#fn("9000r1c0c1q_~332|;" #(#fn(table.foldl)
   #fn("9000r3c0~}|43;" #(#fn(put!))))) #fn(table)) table.invert)
-	%fl:table.keys #185=#fn("9000r1c0c1_|43;" #(#fn(table.foldl)
+	%fl:table.keys #188=#fn("9000r1c0c1_|43;" #(#fn(table.foldl)
 						    #fn("7000r3|g2K;" #())) table.keys)
-	%fl:table.pairs #186=#fn("9000r1c0c1_|43;" #(#fn(table.foldl)
+	%fl:table.pairs #189=#fn("9000r1c0c1_|43;" #(#fn(table.foldl)
 						     #fn("7000r3|}Kg2K;" #())) table.pairs)
-	%fl:table.values #187=#fn("9000r1c0c1_|43;" #(#fn(table.foldl)
+	%fl:table.values #190=#fn("9000r1c0c1_|43;" #(#fn(table.foldl)
 						      #fn("7000r3}g2K;" #())) table.values)
-	%fl:to-proper #188=#fn("8000r1|…40|;|?660|L1;|Me0|N31K;" #(%fl:to-proper) to-proper)
-	%fl:top-level-exception-handler #189=#fn("7000r1c0qe141;" #(#fn("7000r1e0k12c2qc3q41;" #(*stderr*
+	%fl:to-proper #191=#fn("8000r1|…40|;|?660|L1;|Me0|N31K;" #(%fl:to-proper) to-proper)
+	%fl:top-level-exception-handler #192=#fn("7000r1c0qe141;" #(#fn("7000r1e0k12c2qc3q41;" #(*stderr*
   *output-stream* #fn("7000r1c0qc1qt|302;" #(#fn("7000r0e0i20312e1c23041;" #(%fl:print-exception
   %fl:print-stack-trace #fn(stacktrace)))
 					     #fn("7000r1~302c0|41;" #(#fn(raise)))))
   #fn("6000r0~k0;" #(*output-stream*)))) *output-stream*) top-level-exception-handler)
-	%fl:trace #190=#fn("8000r1c0qc1|31312c2;" #(#fn("7000r1c0qc13041;" #(#fn("@000r1e0~317e0c1i10e2c3|c4c5c6c7i10L2|L3L2c8L1c9c7~L2|L3L4L33142;c:;" #(%fl:traced?
+	%fl:trace #193=#fn("8000r1c0qc1|31312c2;" #(#fn("7000r1c0qc13041;" #(#fn("@000r1e0~317e0c1i10e2c3|c4c5c6c7i10L2|L3L2c8L1c9c7~L2|L3L4L33142;c:;" #(%fl:traced?
   #fn(set-top-level-value!) %fl:eval lambda begin write cons quote newline
   apply #{#<unspecified>}#)) #fn(gensym)))
 						    #fn(top-level-value) ok) trace)
-	%fl:traced? #191=#fn("8000r1e0|3116>02c1|31c1~31>;" #(%fl:closure? #fn(function:code)) #(#fn(":000s0c0c1|K312e2302c3|v2;" #(#fn(write)
+	%fl:traced? #194=#fn("8000r1e0|3116>02c1|31c1~31>;" #(%fl:closure? #fn(function:code)) #(#fn(":000s0c0c1|K312e2302c3|v2;" #(#fn(write)
   x %fl:newline #.apply)) ()))
 	%fl:uint16 #fn(uint16) %fl:uint32
 	#fn(uint32) %fl:uint64 #fn(uint64) %fl:uint8
 	#fn(uint8) %fl:ulong #fn(ulong) %fl:uncurry-define
-	#192=#fn("=000r1|NF16D02e0|31F16902e1|31F6P0e2|Me1|31c3c4e5|31e6|3133L341;|;" #(%fl:cadr
+	#195=#fn("=000r1|NF16D02e0|31F16902e1|31F6P0e2|Me1|31c3c4e5|31e6|3133L341;|;" #(%fl:cadr
   %fl:caadr %fl:uncurry-define #fn(list*) lambda %fl:cdadr %fl:cddr) uncurry-define)
-	%fl:untrace #193=#fn("8000r1c0qc1|3141;" #(#fn("9000r1e0|316@0c1~c2|31b2[42;c3;" #(%fl:traced?
+	%fl:untrace #196=#fn("8000r1c0qc1|3141;" #(#fn("9000r1e0|316@0c1~c2|31b2[42;c3;" #(%fl:traced?
   #fn(set-top-level-value!) #fn(function:vals) #{#<unspecified>}#))
 						   #fn(top-level-value)) untrace)
-	%fl:values #194=#fn("9000s0|F16602|NA650|M;~|K;" #() #2#)
-	%fl:vector->list #195=#fn("8000r1c0qc1|31_42;" #(#fn(":000r2a|c0qu2};" #(#fn("8000r1i10~|x[\x7fKo01;" #())))
+	%fl:values #197=#fn("9000s0|F16602|NA650|M;~|K;" #() #2#)
+	%fl:vector->list #198=#fn("8000r1c0qc1|31_42;" #(#fn(":000r2a|c0qu2};" #(#fn("8000r1i10~|x[\x7fKo01;" #())))
 							 #fn(length)) vector->list)
-	%fl:vector.map #196=#fn("8000r2c0qc1}3141;" #(#fn("8000r1c0qc1|3141;" #(#fn(":000r1`~axc0qu2|;" #(#fn(":000r1~|i20i21|[31\\;" #())))
+	%fl:vector.map #199=#fn("8000r2c0qc1}3141;" #(#fn("8000r1c0qc1|3141;" #(#fn(":000r1`~axc0qu2|;" #(#fn(":000r1~|i20i21|[31\\;" #())))
   #fn(vector.alloc))) #fn(length)) vector.map)
-	%fl:void #197=#fn("6000r0c0;" #(#{#<unspecified>}#) void) %fl:wchar
-	#fn(wchar) %fl:zero? #198=#fn("7000r1|`W;" #() zero?) *banner*
+	%fl:void #200=#fn("6000r0c0;" #(#{#<unspecified>}#) void) %fl:wchar
+	#fn(wchar) %fl:zero? #201=#fn("7000r1|`W;" #() zero?) *banner*
 	";  _\n; |_ _ _ |_ _ |  . _ _\n; | (-||||_(_)|__|_)|_)\n;-------------------|----------------------------------------------------------\n\n"
 	*builtins* #(0 0 0 0 0 0 0 0 0 0 0 0 #fn("7000r2|}<;" #())
 		     #fn("7000r2|}=;" #())
@@ -586,79 +596,81 @@
 	#51# cddar #52# cdddar
 	#53# cddddr #54# cdddr
 	#55# cddr #56# char?
-	#57# closure? #58# compile
-	#59# compile-and #60# compile-app
-	#61# compile-arglist #62# compile-begin
-	#63# compile-builtin-call #64# compile-f
-	#65# compile-f- #66# compile-for
-	#67# compile-if #68# compile-in
-	#69# compile-or #70# compile-prog1
-	#71# compile-short-circuit #72# compile-sym
-	#73# compile-thunk #74#
+	#57# closure? #58# compare<=
+	#59# compile #60# compile-and
+	#61# compile-app #62# compile-app*
+	#63# compile-arglist #64# compile-begin
+	#65# compile-builtin-call #66# compile-f
+	#67# compile-f- #68# compile-for
+	#69# compile-if #70# compile-in
+	#71# compile-or #72# compile-prog1
+	#73# compile-short-circuit #74# compile-sym
+	#75# compile-thunk #76#
 	compile-unknown-call #fn("6000r2^;" #() compile-unknown-call)
-	compile-while #75# const-to-idx-vec
-	#76# copy-tree #77# count
-	#78# define-head-name #79# delete-duplicates
-	#80# disassemble #81# div
-	#82# emit #83#
-	emit-optional-arg-inits #84# encode-byte-code
-	#85# error #86# eval
-	#87# even? #88# every
-	#89# expand #90# expand-define
-	#91# expand-macro-call #92# filter
-	#93# fits-i8 #94# foldl
-	#95# foldr #96# for-each
-	#97# function-source #98# get-defined-vars
-	#99# hex5 #100# identity
-	#101# in-env? #102# index-of
-	#103# io.readall #104# io.readline
-	#105# io.readlines #106# iota
-	#107# keyword->symbol #108# keyword-arg?
-	#109# lambda-arg-names #110# lambda-vars
-	#111# last-pair #112# lastcdr
-	#113# length= #114# length>
-	#115# list->vector #116# list-head
-	#117# list-ref #118# list-tail
-	#119# list? #120# load
-	#121# load-process #122# lookup-sym
-	#123# macrocall? #124# macroexpand-1
-	#125# make-code-emitter #126# make-label
-	#127# make-perfect-hash-table #128# make-system-image
-	#129# map! #130# map-int
-	#131# mark-label #132# max
-	#133# member #134# memv
-	#135# min #136# mod
-	#137# mod0 #138# nan?
-	#139# nary-compare #140# nary<
-	#141# nary= #142# negative?
-	#143# nestlist #144# newline
-	#145# nnn #146# nreconc
-	#147# odd? #148# positive?
-	#149# princ #150# print
-	#151# print-exception #152# print-stack-trace
-	#153# print-to-string #154# printable?
-	#155# proper-part #156# quote-value
-	#157# random #158# read-all
-	#159# read-all-of #160# ref-int16-LE
-	#161# ref-int32-LE #162# repl
-	#163# resolve-global #fn("6000r1|;" #()) revappend
-	#164# reverse #165# reverse!
-	#166# reverse!- #167# reverse-
-	#168# self-evaluating? #169# separate
-	#170# set-syntax! #171# simple-sort
-	#172# special-forms (quote if begin prog1 lambda and or while for
-			     return set! define trycatch)
-	splice-form? #173# string.join
-	#174# string.lpad #175# string.map
-	#176# string.rep #177# string.rpad
-	#178# string.tail #179# string.trim
-	#180# symbol-syntax #181# table.clone
-	#182# table.foreach #183# table.invert
-	#184# table.keys #185# table.pairs
-	#186# table.values #187# to-proper
-	#188# top-level-exception-handler
-	#189# trace #190# traced?
-	#191# uncurry-define #192# untrace
-	#193# values #194# vector->list
-	#195# vector.map #196# void
-	#197# zero? #198#)
+	compile-while #77# const-to-idx-vec
+	#78# copy-tree #79# count
+	#80# define-head-name #81# delete-duplicates
+	#82# disassemble #83# div
+	#84# emit #85#
+	emit-optional-arg-inits #86# encode-byte-code
+	#87# error #88# eval
+	#89# even? #90# every
+	#91# expand #92# expand-define
+	#93# expand-macro-call #94# filter
+	#95# fits-i8 #96# foldl
+	#97# foldr #98# for-each
+	#99# function-source #100# get-defined-vars
+	#101# hex5 #102# identity
+	#103# in-env? #104# index-of
+	#105# io.readall #106# io.readline
+	#107# io.readlines #108# iota
+	#109# keyword->symbol #110# keyword-arg?
+	#111# lambda-arg-names #112# lambda-vars
+	#113# last-pair #114# lastcdr
+	#115# length= #116# length>
+	#117# list->vector #118# list-head
+	#119# list-ref #120# list-tail
+	#121# list? #122# load
+	#123# load-process #124# lookup-sym
+	#125# macrocall? #126# macroexpand-1
+	#127# make-code-emitter #128# make-label
+	#129# make-perfect-hash-table #130# make-system-image
+	#131# map! #132# map-int
+	#133# mark-label #134# max
+	#135# member #136# memv
+	#137# min #138# mod
+	#139# mod0 #140# nan?
+	#141# nary-compare #142# nary<
+	#143# nary= #144# negative?
+	#145# nestlist #146# newline
+	#147# nnn #148# nreconc
+	#149# odd? #150# positive?
+	#151# princ #152# print
+	#153# print-exception #154# print-stack-trace
+	#155# print-to-string #156# printable?
+	#157# proper-part #158# quote-value
+	#159# random #160# read-all
+	#161# read-all-of #162# ref-int16-LE
+	#163# ref-int32-LE #164# repl
+	#165# resolve-global #fn("6000r1|;" #()) revappend
+	#166# reverse #167# reverse!
+	#168# reverse!- #169# reverse-
+	#170# self-evaluating? #171# separate
+	#172# set-syntax! #173# simple-compare
+	#174# simple-sort #175# special-forms
+	(quote if begin prog1 lambda and or while for return set! define
+	 trycatch)
+	splice-form? #176# string.join
+	#177# string.lpad #178# string.map
+	#179# string.rep #180# string.rpad
+	#181# string.tail #182# string.trim
+	#183# symbol-syntax #184# table.clone
+	#185# table.foreach #186# table.invert
+	#187# table.keys #188# table.pairs
+	#189# table.values #190# to-proper
+	#191# top-level-exception-handler
+	#192# trace #193# traced?
+	#194# uncurry-define #195# untrace
+	#196# values #197# vector->list
+	#198# vector.map #199# void
+	#200# zero? #201#)
diff --git a/system.lsp b/system.lsp
index 1fad948..3d56921 100644
--- a/system.lsp
+++ b/system.lsp
@@ -149,10 +149,10 @@
 
 (define (>  a b) (< b a))
 (define (nan? x) (or (equal? x +nan.0) (equal? x -nan.0)))
-(define (<= a b) (not (or (< b a)
-                          (and (nan? a) (nan? b)))))
-(define (>= a b) (not (or (< a b)
-                          (and (nan? a) (nan? b)))))
+; TeXmacs: the order of < (any values), but false when a number is a NaN
+; ((<= 1 +nan.0) was true); compile-app compiles them so
+(define (<= a b) (if (< b a) #f (if (number? a) (if (< a b) #t (= a b)) #t)))
+(define (>= a b) (<= b a))
 (define (negative? x) (< x 0))
 (define (zero? x)     (= x 0))
 (define (positive? x) (> x 0))
diff --git a/tests/unittest.lsp b/tests/unittest.lsp
index fbe40a2..bee157b 100644
--- a/tests/unittest.lsp
+++ b/tests/unittest.lsp
@@ -95,6 +95,13 @@
 (assert (equal? (< 3 +nan.0) (> +nan.0 (double 3))))
 (assert (equal? (> 3 +nan.0) (> (double 3) +nan.0)))
 (assert (not (>= +nan.0 +nan.0)))
+; TeXmacs: false with one NaN too, also when compiled as instructions
+(assert (not (<= 1 +nan.0)))
+(assert (not (<= +nan.0 1)))
+(assert (not (>= 1 +nan.0)))
+(assert (not ((lambda (a b) (<= a b)) 1 +nan.0)))
+(assert ((lambda (a b) (and (<= a b) (>= b a) (> b a) (<= a a))) 1 2))
+(assert ((lambda (a b) (<= a b)) "a" "a"))
 
 ; comparing strings
 (assert (< "a" "b"))
-- 
2.50.1 (Apple Git-155)

